### PR TITLE
Add mark as read option in room setting

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes https://github.com/vector-im/element-web/issues/24053

Screenshot 2022-12-19 at 16 04 17

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add mark as read option in room setting ([\#9798](https://github.com/matrix-org/matrix-react-sdk/pull/9798)). Fixes vector-im/element-web#24053.

### CODE REVIEW

File: res/css/views/context_menus/_RoomGeneralContextMenu.pcss [hunk 1]
```diff
@@ -6,6 +6,10 @@
     mask-image: url("$(res)/img/element-icons/roomlist/low-priority.svg");
 }
 
+.mx_RoomGeneralContextMenu_iconMarkAsRead::before {
```
Comment (gsouquet) [inline res/css/views/context_menus/_RoomGeneralContextMenu.pcss:9]: This should be done using the SVG icon approach but would warrant too large of a refactor on an untested part of the codebase that I do not want to test right now.

File: src/utils/notifications.ts [hunk 2]
```diff
@@ -59,27 +59,61 @@ export function localNotificationsAreSilenced(cli: MatrixClient): boolean {
...
+    const roomEvents = room.getLiveTimeline().getEvents();
+    const lastThreadEvents = room.lastThread?.events;
+
+    const lastRoomEvent = roomEvents?.[roomEvents?.length - 1];
+    const lastThreadLastEvent = lastThreadEvents?.[lastThreadEvents?.length - 1];
+
+    const lastEvent =
+        (lastRoomEvent?.getTs() ?? 0) > (lastThreadLastEvent?.getTs() ?? 0) ? lastRoomEvent : lastThreadLastEvent;
+
+    let promise: Promise<{}>;
```
Comment (florianduros) [inline src/utils/notifications.ts:78]: This variable seems to be unused

File: src/utils/notifications.ts [hunk 3]
```diff
@@ -59,27 +59,61 @@ export function localNotificationsAreSilenced(cli: MatrixClient): boolean {
...
+        // notified a user.
+        room.setUnreadNotificationCount(NotificationCountType.Highlight, 0);
+        room.setUnreadNotificationCount(NotificationCountType.Total, 0);
+        for (const thread of room.getThreads()) {
+            room.setThreadUnreadNotificationCount(thread.id, NotificationCountType.Highlight, 0);
+            room.setThreadUnreadNotificationCount(thread.id, NotificationCountType.Total, 0);
+        }
+    }
+
+    return promise ?? {};
```
Comment (florianduros) [inline src/utils/notifications.ts:104]: If `promise` is unused, `return Promise.resolve()` should work here

**Review by florianduros — APPROVED** [review]
LGTM, small comments

**Comment by gsouquet** [comment]
@DoM1niC yes and no.
Ideally I agree with you we wouldn't want to have that! However we have some stuck notifications and this button can really help

**Pull request merged**